Add docs for blending

almarklein opened the pull request:
Ref discussion in #1120.

I refactored the current guide, moving the content to multiple files, to make room to go into some topics deeper, like we do here for transparency.

The text does not describe how Pygfx currently work, but how I think it should work. Open for suggestions. The idea is to get (more or less) agreement on this, then implement it accordingly.

Korijn commented on docs/blending.rst:
```diff
@@ -0,0 +1,261 @@
+* Objects are first sorted by their ``object.render_order``. Users should generally not need this, but it allows full control over the render order.
+* Then bases on their ``material.alpha mode`` and ``material.depth_write``, in order: "opaque" and "dither", "blend" with depth write, "blend", "weighted".
+* The object's distance to the camera. Note that this won't help all cases, since objects
+  can still intersect other objects (and themselves).
```
```suggestion
  can still intersect other objects (and themselves). This sorting step is intended primarily to reduce overdraw.
```

Korijn commented:
Focusing on the API design: it looks fine to me. I'm curious to hear from @panxinmiao. Are there any overlooked use cases?

Korijn commented on docs/blending.rst:
```diff
@@ -0,0 +1,261 @@
+on three factors:
+
+* Objects are first sorted by their ``object.render_order``. Users should generally not need this, but it allows full control over the render order.
+* Then bases on their ``material.alpha mode`` and ``material.depth_write``, in order: "opaque" and "dither", "blend" with depth write, "blend", "weighted".
```
Then I would propose this formatting change to make it easier to parse the order

panxinmiao commented:
I'm not entirely sure if I expressed my point clearly. If my wording caused any confusion, I sincerely apologize.

I would like to explain my concerns again:

### 1、`alpha_mode` conflates object classification semantics and pipeline configuration semantics, leading to conceptual confusion.

Values like `alpha_mode="blend"`, alongside opaque, weighted, and dither, essentially serve as a classification system indicating how an object should be rendered. This is a high-level, engine-side logic.
On the other hand, values like "add" or "multiply" under blend_mode (or sometimes directly under alpha_mode) represent specific GPU-level blending configurations. These define how colors are computed in the rendering pipeline and belong to the low-level graphics API domain.

Mixing these two kinds of semantics into a single enum or parameter tree leads to confusion. And, blending settings in the rendering pipeline are not exclusive to transparent objects, using “whether blending is enabled” as the criterion for determining if an object is transparent is problematic and misleading.

For example:
```
m.alpha_mode = "add"
```
This raises the question:
- Does this mean the object is a transparent additive blend, requiring back-to-front sorting?
- Or is it an opaque object using additive blending, rendered in a forward (unsorted) opaque pass?


### 2、The current presets under `alpha_mode="blend"`—such as "over", "add", "subtract", and "multiply"—are too limited.
They do not adequately cover the full range of blending configurations supported by modern rendering pipelines. 
For instance:

- Min/Max blending
- Custom src/dst blend factors
- Blend constants

panxinmiao commented on docs/blending.rst:
```diff
@@ -0,0 +1,250 @@
+
+The renderer sorts objects based on three factors:
+
+* Objects are first sorted by their ``object.render_order``. Users should generally not need this, but it allows full control over the render order.
+* Then based on ``material.alpha mode`` and ``material.depth_write``, in order:
```
I think 'render_order' property can only control the rendering order within a specific rendering classification. In other words, an object's classification (e.g., opaque, alpha-blended, weighted) is the primary factor in determining render order, and `render_order` only applies within that classification group.

This is because, fundamentally, different classifications correspond to different render passes, which are executed in separate phases. Therefore, cross-pass sorting is not feasible and generally makes no sense. `render_order` is only meaningful for controlling the order of objects within the same render pass.

almarklein replied:
This is a great point. I was actually thinking about this yesterday. I agree that often you want to offset the object render order *within a group* of objects (opaque/blended). However, I think there are also cases where the user wants to re-order beyond groups, one example being objects in the weighted and blended class; the renderer renders weighted last, but not all use-cases may want this.

Maybe we can distinguish two super-groups, within which shifts makes sense. One for objects producing opaque fragments (opaque and dither alpha modes), and one blended for the rest.

panxinmiao replied:
The core issue here is that sorting across different render passes is difficult to implement or can introduce significant overhead, while offering very little benefit in most practical scenarios.

Regarding the example you mentioned, I have also added comments in the renderer (see #974) to explain the behavior when handling weighted blending objects.

To summarize: it is generally not advisable to mix alpha blending and weighted blending objects within the same scene, because they represent fundamentally different transparency strategies. A single scene should adopt a consistent approach to transparency—either order-dependent alpha blending or OIT via weighted blending. Moreover, weighted blending relies on all participating objects contributing to the same compositing process.
Even if, for some reason, different transparency strategies are used in the same scene, correct rendering is not guaranteed. This is because weighted blending objects do not participate in depth sorting—which is, in fact, one of the main reasons to use weighted blending (OIT) in the first place.

I understand you might argue that users can still rely on `render_order` to carefully control the draw order when interleaving weighted and alpha-blended objects.(But if so, what’s the point of using OIT in the first place?). Weighted and alpha blending require different render passes. Interleaving them would mean switching between render passes multiple times during a single frame, which is extremely costly. Modern GPUs are explicitly designed to minimize render pass switches for performance reasons.

In general, from a renderer’s perspective, the number and order of render passes should be determined by the engine’s rendering strategy—not dynamically changed based on object arrangement in the scene.

The reason we categorize objects into different groups is precisely because they use different render passes and must be rendered in different phases of the pipeline. All objects in a scene should be clearly grouped and rendered during well-defined stages—not arbitrarily interleaved based on per-object sorting.

If users have special needs where an object must be rendered out-of-order (e.g., a transparent object rendered between opaque layers), they should render it in a separate scene, with explicit control over rendering order and timing.

almarklein replied:
The argument about switching pipelines makes sense if all objects were of the same type (and to some degree the same material), but the use cases we're dealing with is e.g. stiching multiple images using 'weighted' and then rendering semi-transparent points on top. The pipelines are not the same anyway, so switching order will not affect performance.

> To summarize: it is generally not advisable to mix alpha blending and weighted blending objects within the same scene, because they represent fundamentally different transparency strategies.

I see no objection why this should not be possible 🤷  (except for this sorting issue).

In any case, I shall address the original issue.

panxinmiao replied:
> The argument about switching pipelines makes sense if all objects were of the same type (and to some degree the same material), but the use cases we're dealing with is e.g. stiching multiple images using 'weighted' and then rendering semi-transparent points on top. The pipelines are not the same anyway, so switching order will not affect performance.

**the number and order of render passes should be determined by the engine’s rendering strategy—not dynamically changed based on object arrangement in the scene.**

In our current grouping design (as implemented in #974), we have three fixed rendering passes: the opaque pass, the transparent pass (which may use two passes render twice), and the weighted blending pass. **The number and order of these render passes are fixed**.

In an extreme case, if there are 100 alpha-blended objects and 100 weighted-blending objects interleaved with each other, does that mean we would need to use 200 render passes if we strictly follow their `render_order`?

almarklein commented:
> No, it's important. This is exactly where the problem lies, we need clear and unambiguous categorization.

Please help me understand, by providing an example where an object has `transparent=False`, which cannot be derived from `depth_write`. Or maybe another way to phrase it, you seem to suggest that an object can be `transparent=False, alpha_mode='blend', depth_write=False`. How do you expect the renderer to handle that?

panxinmiao commented:
> > No, it's important. This is exactly where the problem lies, we need clear and unambiguous categorization.
> 
> Please help me understand, by providing an example where an object has `transparent=False`, which cannot be derived from `depth_write`. Or maybe another way to phrase it, you seem to suggest that an object can be `transparent=False, alpha_mode='blend', depth_write=False`. How do you expect the renderer to handle that?

For example, scene backgrounds (or objects meant to serve as scene backgrounds), decals, or helper objects and markers for debugging or visualization purposes. I once had a use case where an object had a hidden proxy geometry—only the proxy geometry wrote to the depth buffer, while the visible object itself only wrote to the color buffer but not to depth

almarklein commented:
Yeah but how do you think the renderer should use the value of `.transparent`? From what I understand, objects with `transparent=True` are rendered first, but for the case of backgrounds this is not enough, because you want to render before all the other opaque objects. I'm not sure how the decals etc. fit in. Can you elaborate?

almarklein commented:
Thanks for the explanation; I understand your point better now. It has everything to do with controlling the order that objects are rendered in. I'll give this some more thought ...

almarklein commented:
I'm moving rather slow since I am on holiday and try to get some Pygfx done in the late evenings / early mornings.

I've updated this pr and the #1144. I've not included a `transparency` property, because from the use-cases put forward it seems mostly used as a trick to force things to be rendered earlier or later than usual. Instead I added `object.render_group` which allows controlling multiple "layers", as if using multiple calls to `render()` except defined in the scene graph, which covers use-cases such as backgrounds and overlays. 

I went a bit back and forth with naming things but I settled on:
* `alpa_mode`: a string preset for the majority of cases (and to which we can add if we want).
* `alpha_config`: a dict that fully describes the alpha handling in one place (instead of separate dicts for composite, dither, etc. which felt awkward).
* `alpha_method`: represents the 'group': either "opaque", "composite", "stochastic"  or "weighted".
* Note that the names of alpha-method are more technical, while the names in alpha-mode are a bit more user-friendly, e.g. "solid" and "blend" (for the over-operator).

Further:
* The renderer employs an opaque and transparency pass, like most render engines do. Which pass to assign an object to is based on the alpha-method, but users can override this in the `alpha_config`. These cases should be really rare.
* Objects with alpha-method 'weighted' are grouped together because of the way the z-sorting happens. Though they can still be "grouped" using `ob.render_order`.
* I added alpha-mode 'auto' as a somewhat plug-n-play mode, especially for 2D-ish scenes. I'm not entirely happy about it, but I think it more or less represents the earlier behavior.

almarklein commented:
I updated this pr and #1144. IMO the ``render_queue`` idea solves the whole render-order issue in an elegant way that can also be super-flexible for special use-cases. I hope we can start to wrap this up ...

Korijn commented:
> I updated this pr and #1144. IMO the `render_queue` idea solves the whole render-order issue in an elegant way that can also be super-flexible for special use-cases. I hope we can start to wrap this up ...

To me this looks fantastic but I also think you should just move on and accept the outcome. In the interest of other priorities. :)

panxinmiao commented:
I believe there is one more point that needs to be clarified and documented.

Specifically, which rendering passes/stages our renderer internally uses (currently, we have three passes/stages in #974: Opaque, Transparent, and Weighted),

and according to our API design, what rules govern the assignment of objects to these specific rendering passes/stages.

Additionally, is `render_queue` intended to be user-configurable? Since the different rendering stages are inherent internal passes of the renderer and cannot be reordered arbitrarily, could this conflict with `render_queue`?

In other words, if an object is assigned to the Transparent stage based on the above rules, but the user sets its `render_queue` to 2000, what would happen?

panxinmiao commented:
Thank you for the detailed explanation. Now I have a better understanding of the API.

almarklein commented:
> I'm going to answer the questions, but they are based on my understanding of the diff. Correct me if I get anything wrong!

I agree with all the answers 👍  The `render_queue` is an alternative to the binary (opaque vs transparent) render-pass concept that ThreeJS uses. It solves cases like background and overlays, allows precise insertions, custom passes, and is explicit.

almarklein commented:
> I would love to merge this!

Should ideally happen together with / right after #1144

Korijn commented on docs/transparency.rst:
```diff
@@ -0,0 +1,421 @@
+the case; they can e.g. be used as the reference value in alpha testing.
+
+In any case, the alpha value represents a weight for how the object is combined with
+other objects, and it's applicance is fully defined by ``material.alpha_config``.
```
```suggestion
other objects, and its application is fully defined by ``material.alpha_config``.
```

Korijn commented on docs/transparency.rst:
```diff
@@ -0,0 +1,421 @@
+2. Set ``material.alpha_mode`` to a preset string:
+
+These provide configurations for common cases. Examples are "solid",
+"blend", "dither", and several more.
```
```suggestion
"blend", "dither", and several more. See below for a full list of options and descriptions.
```

Korijn commented on docs/transparency.rst:
```diff
@@ -0,0 +1,421 @@
+* "auto": classic alpha blending, with ``depth_write`` defaulting to True if ``.opacity==1``.
+* "blend": classic alpha blending using the over-operator.
+* "add": additive blending that adds the fragment color, multiplied by alpha.
+* "subtract": subtractuve blending that removes the fragment color.
```
```suggestion
* "subtract": subtractive blending that removes the fragment color.
```

Korijn commented on docs/transparency.rst:
```diff
@@ -0,0 +1,421 @@
+
+The ``material.render_queue`` is an integer that represents the group that the renderer uses to sort objects.
+The property is intended for advanced use; it is determined automatically
+based on ``alpha_method``, ``depth_write`` and ``alpha_test``. It's values lays between 1 and 4999,
```
```suggestion
based on ``alpha_method``, ``depth_write`` and ``alpha_test``. Its values lays between 1 and 4999,
```

Korijn commented on docs/transparency.rst:
```diff
@@ -0,0 +1,421 @@
+
+* The ``material.render_queue``.
+* The ``object.render_order``.
+* The object's distance to the camera, either front-to-back or back-to-front, depending on the ``render_queue``. Weighted objects are not sorted.
```
What are weighted objects?

Korijn commented on docs/transparency.rst:
```diff
@@ -0,0 +1,421 @@
+    .. code-block:: js
+
+        // ThreeJS
+        m.transparent.false;  // default
```
```suggestion
        m.transparent = false;  // default```

Korijn commented on docs/transparency.rst:
```diff
@@ -0,0 +1,421 @@
+    .. code-block:: py
+
+        # Pygfx
+        # (the object gets automatically rendered at the very start of the transparency-pass)
```
because of render queues right? or... why?

almarklein replied:
yeah, I updated this comment.

Korijn commented:
I figured it would be worthwhile to give this a linguistic review and to check for clarity :) if you can address my comments we can merge this.

Korijn commented on docs/transparency.rst:
```diff
@@ -0,0 +1,421 @@
+    .. code-block:: py
+
+        # Pygfx
+        ob.material.render_queue = 1000
```
```suggestion
        ob.material.render_queue = 1000  # the render queue for backgrounds
```

Korijn commented on docs/transparency.rst:
```diff
@@ -0,0 +1,421 @@
+Render queue
+------------
+
+The ``material.render_queue`` is an integer that represents the group that the renderer uses to sort objects.
```
Maybe also mention here that it is an integer and not an Enum because that would prevent custom assignment? I can already picture the PRs/issues about this.

almarklein commented on docs/transparency.rst:
```diff
@@ -0,0 +1,421 @@
+    .. code-block:: js
+
+        // ThreeJS
+        // (put at the beginning of the opaque-pass, so no blending possible.)
```
I'll remove the comment about no blending possible, because it's confusing. But its a limitation of ThreeJS: blending is only enabled if you use `transparent=false`.

Korijn commented on docs/transparency.rst:
```diff
@@ -0,0 +1,421 @@
+
+In contrast, the "composite" and "weighted" methods result in semi-transparent fragments,
+and by default have ``depth_write=False``. The renderer sorts these object back-to-front to
+improve the chance of correct blending. Note that the 'auto' mode is an exception to this rule.
```
@almarklein you missed this comment I think!

almarklein replied:
Woops! Yes I did. Addressed it by making explicit what I meant.

Korijn commented on docs/transparency.rst:
```diff
@@ -124,13 +124,14 @@ Method "weighted" (order independent blending):
 In contrast, the "composite" and "weighted" methods result in semi-transparent fragments,
 and by default have ``depth_write=False``. The renderer sorts these object back-to-front to
-improve the chance of correct blending. Note that the 'auto' mode is an exception to this rule.
+improve the chance of correct blending. Note that with the 'auto' mode, de default ``depth_write`` depends
```
```suggestion
improve the chance of correct blending. Note that with the 'auto' mode, the default ``depth_write`` depends
```

Korijn commented:
I guess it was a little frustrating process, but I am really impressed by the outcome. ❤️

hmaarrfk commented:
uge

panxinmiao commented:
Sorry to bring up this topic again, 😓

After several days of hands-on practice and investigation, I can confirm that the renderQueue here is not the object classification mechanism we need.

Essentially, it is a higher-level sorting mechanism. On the CPU side, it buckets and sorts all objects based on their renderQueue to determine the order and grouping strategy of DrawCalls.  
Although it ultimately determines which rendering phase/Pass an object belongs to in the renderer, it functions more like an internal identifier (even though it can be directly specified by users).  
It still doesn’t solve the fundamental question: “How are objects classified (i.e., assigned to a specific phase) by default?”  
Even in Unity, users are required to set the Queue or RenderType tag in the material shader to determine the final RenderQueue.

- "Opaque" → 2000  
- "TransparentCutout" → 2450  
- "Transparent" → 3000  

The only difference is that Unity allows users to directly set the RenderQueue.

However, Queue or RenderType itself is explicit, unambiguous, and requires user input—the engine does not infer it (and in my opinion, cannot infer it, as it is essentially a meta-attribute that reflects user intent).  
If not set, the engine simply assigns a default value (e.g., Queue=2000) without any additional inference.

The Queue or RenderType values—Opaque, TransparentCutout, and Transparent—are exactly the internal phases and object classifications we need in the renderer.  
Just like the `transparent` property in three.js or the `TransparencyMode` property in Babylon.js, these are meta-attributes that determine user intent and directly assign objects to the appropriate rendering phase. I believe the engine should not attempt to infer them (and in fact, cannot reliably do so).

_Additionally, while reviewing Babylon.js documentation, I noticed that `TransparencyMode` can be left empty, and Babylon.js uses a very complex (though still explicit) logic to automatically infer it—unlike other engines that simply assign a default value. I was initially curious why Babylon.js adopted this approach, but after some research, I believe it’s merely a historical burden.
As mentioned in https://github.com/BabylonJS/Babylon.js/pull/16144, this might be removed in the future._

---

Our current issue is that although `alpha_method` seems usable as a classification basis, it is an internal value that cannot be set directly or independently, and it’s opaque to users.  
Even if it can be configured via the `alpha_config` dictionary, `alpha_method` is tied to the GPU pipeline’s blending settings and cannot serve as an independent intent-based identifier for grouping or phase assignment. This brings us back to square one. I’ve expressed this viewpoint multiple times that internal renderer phase division and GPU blending configuration are orthogonal concepts and should not be forcibly coupled.

I’d like to reiterate: object classification and renderer phase division are based on the engine’s own scene rendering logic and design philosophy—they exist to enable specific engine capabilities, e.g., supporting Weighted Blend OIT objects by introducing a dedicated Weighted phase (the rationale for treating it as a separate phase can be discussed elsewhere).  
Most engines classify objects into at least opaque and transparent categories (with corresponding internal render phases) because, in most scenarios, opaque and transparent objects require different processing logic. Moreover, between rendering opaque and transparent objects, the renderer may need to perform some certain tasks (e.g., generating SSR-based transmissive light sampling maps for physically based transparency).

GPU blending, on the other hand, is a rendering pipeline capability with various applications. It is not synonymous with transparency or alpha blending, nor is it exclusive to a specific object category or render phase.

A simple example:  
Even for opaque objects (from the renderer’s perspective, meaning they are rendered in the opaque phase—not necessarily visually opaque), users can leverage blending. A common use case is using additive blending for color accumulation.

Furthermore, if desired, users can achieve correct alpha-blended transparency effects even in this phase.  
(Assuming source colors are pre-multiplied by α for simplicity)  

When blending back-to-front iteratively (where `dst` is the accumulated result and `src` is the current fragment), the standard blend equation is:  
  `dst ← src + (1 - src_α) * dst`, 
  configured with:  
  `SrcFactor = ONE`, `DstFactor = ONE_MINUS_SRC_ALPHA`.  

When blending front-to-back (typical for the opaque phase), use:  
  `dst ← dst + (1 - dst_α) * src`,
  configured with:  
  `SrcFactor = ONE_MINUS_DST_ALPHA`, `DstFactor = ONE`.  

The two are mathematically equivalent.  

However, if users do this, the renderer still treats it as an opaque object and processes it in the opaque phase. This means all engine logic after the opaque phase (e.g., generating SSR buffers for physically based transparency, special handling for transparent-phase objects like dual-side dual-pass rendering, etc.) occurs after this object is rendered.

---

At last, I feel the current design of concepts like `alpha_mode`, `alpha_config`, and `alpha_method` is overly complex—even for someone relatively familiar with the API like myself, it’s easy to get confused.

Additionally, while well-intentioned to minimize user cognitive load (by reducing the number of values users need to know and set), having `auto` behaviors for many enums may actually cause more confusion. The engine needs to know the user’s explicit intent, and users working directly with the engine should understand its characteristics. Higher-level applications built on top of the engine should be responsible for providing abstractions based on specific use cases.

This is similar to how developers using pygfx must understand what Geometry represents, what Material represents, and what vertex positions in Geometry mean—while end-users of applications built on pygfx don’t need to know these details.

For example, the current code defaults `alpha_mode='auto'` but uses `"blend"` as the fallback behavior. I understand the reasoning: even for opaque objects, using blending won’t cause issues (since alpha=1), ensuring broad compatibility. However, this causes all default objects to be classified as transparent (renderQueue 3000) and assigned to the transparent phase (using the Transparent Pass), leading to unexpected results. It took me about an hour to figure out the reason why my rendering result was incorrect.

I think, trying to “do everything for the user” in a rendering engine often leads to bigger problems.

almarklein commented:
I have a feeling the confusion comes from a need to want to classify an object as opaque or transparent.

The short answer is this:
```py
matrerial.render_queue = 2000  # == material.transparent = False
material.render_queue = 3000  # == material.transparent = True
```

Yes, the value of `render_queue` is automatically derived from `alpha_method`, `depth_write`, and whether alpha test is used. But as far as "intent" goes, users should probably just set `render_queue` directly.

> Even if [`alpha_method`] can be configured via the `alpha_config` dictionary, `alpha_method` is tied to the GPU pipeline’s blending settings and cannot serve as an independent intent-based identifier for grouping or phase assignment. This brings us back to square one.

You are correct that alpha-method is a poor identifier for intent by itself. You should probably set `material.render_queue` instead.


> A simple example:
Even for opaque objects (from the renderer’s perspective, meaning they are rendered in the opaque phase—not necessarily visually opaque), users can leverage blending.
> [...]
> However, if users do this, the renderer still treats it as an opaque object and processes it in the opaque phase. 

Same answer, set `render_queue`.

> At last, I feel the current design of concepts like `alpha_mode`, `alpha_config`, and `alpha_method` is overly complex—even for someone relatively familiar with the API like myself, it’s easy to get confused.

I agree it's a rather complex system. I have struggled with this a lot, and it still frustrates me. I would very much present users with a system that Just Works. I actually think we could get close if we adopt techniques such as adaptive transparency, mlab etc. which insert output fragments into a multi-layer render target. But such techniques suffer a significant performance and memory hit. So I've somewhat accepted that we cannot really solve this problem for the user. Instead we offer more solutions that most engines do.

So the complexity comes partly from the fact that we have included support for stochastic and weighted methods. This offers more ways to deal with transparency, at the cost of a  more complex API 🤷 

> having auto behaviors for many enums may actually cause more confusion.

Also true. I can then recommend to explicitly set most such enums. I think it makes sense to recommend users to use explicit values in the docs.

BTW: the 'auto' alpha-mode also comes from the fact that we have multiple objects (lines, points, text) that have semi-transparent fragments for aa even if the object is opaque. Such objects can still be rendered fine, especially in 2D scenario's.

almarklein commented:
There is no "opaque rendering stage" or "transparent rendering stage". And there is no way to classify an object as either opaque or transparent.

There's two main things:
* How is the alpha value used, e.g. blended, dithered, ignored. This is defined by `alpha_config`. Most users will set it via the `alpha_mode` presets.
* In what order the objects are rendered. This is defined by `render_queue`. It's default value is derived by guessing intent from alpha mode and depth_write, covering common cases. Setting the value manually, gives users *a lot* of flexibility in controlling object order.

There is also `ob.render_order` to control an object's order within a render queue. In ThreeJS this is used in certain cases to push objects to the end/beginning of the current pass. With the concept of the `render_queue` user will need it less often.

> For example, when the opacity=1, these objects should be classified as opaque, with no reason to place them in the transparent rendering stage.

To address this point specifically as an example: objects that are blended but also write depth, by default end up in render_queue 2600, which sits somewhere in between the default queues for opaque (2000) and transparent (3000) objects, and objects are sorted back-to-front (so that any aa edges blend correctly).

panxinmiao commented:
> There is no "opaque rendering stage" or "transparent rendering stage". And there is no way to classify an object as either opaque or transparent.

Dividing the rendering process into stages is a necessary prerequisite for implementing some modern real-time rendering techniques. I am certain that almost all modern rendering engines at least separate the opaque and transparent stages, since many techniques require the renderer to perform specific operations between these two stages—for example, generating the SSR buffer. This was also the original motivation behind introducing object classification and stage division in #974.

I will explain in detail in #974 when I have time.

panxinmiao commented:
> BTW: the 'auto' alpha-mode also comes from the fact that we have multiple objects (lines, points, text) that have semi-transparent fragments for aa even if the object is opaque. Such objects can still be rendered fine, especially in 2D scenario's.

I’d also like to add a note here: as I understand it, the current "aa" for certain opaque objects (such as points, lines, and text) relies on transparent fragments, which makes the API design and implementation somewhat awkward.
In my view, the reason for adopting this fragment-based alpha blending approach to “aa” is largely due to the lack of a proper implementation of PMA (premultipliedAlpha). If I’m not mistaken, this “aa” is mainly intended to mitigate edge artifacts and black borders, but the same can be achieved correctly with PMA. This would make the notion of “opaque” more consistent across API semantics, visual results, and internal implementation. Combined with MSAA and other techniques, it can further improve the overall anti-aliasing quality.
My plan is to systematically investigate and refine the logic of premultipliedAlpha after resolving the ColorSpace-related issues, in order to establish a complete and correct workflow for textures and color management. Until then, including in changes like #974, I will aim to maintain compatibility and preserve the current “aa” behavior.

panxinmiao commented:
> But I'm pretty sure this can be made to work without dividing objects into 4 render stages (opaque, transmissive, transparent, weighted)...
> 

The rendering process is currently divided into three stages: opaque, transparent, and weighted blend. Transmissive and transparent objects are handled within the same rendering stage/render pass. However, I deliberately sort transmissive objects before regular transparent ones using renderQueue. This ordering brings two benefits, as transmissive objects typically do not rely on GPU blending and usually `depth_write=True` (from the perspective of GPU rendering pipeline, it behaves like an "opaque" object), :

- When `depth_write = True`, it allows better utilization of early-Z testing.

- When users intentionally set its `depth_write = False`, it can, to some extent (though not perfectly), simulate the visibility of regular transparent objects behind them.

> The three render stages representing the generally-transparent objects, are forced in the order: transmissive, transparent, weighted. Ignoring the `render_queue` set by the user.

This is exactly the issue I mentioned earlier: renderQueue is essentially a higher-level sorting mechanism, while the division of rendering stages is an inherent part of the renderer’s logic. No matter how renderQueue is adjusted, the sorting only takes effect within a given rendering stage, and cannot cross stage boundaries.
That said, in the current implementation, the distinction between opaque and transparent stages is determined by whether renderQueue < 2500. As a result, apart from weighted blend objects, the rendering order can be regarded as fully following the user-defined renderQueue; weighted blend objects are the only exception.

---

Before considering how to handle Weighted Blended objects, I attempted to research the implementations of current mainstream rendering engines. However, I found that the vast majority of engines do not have built-in OIT (Order-Independent Transparency) pipelines. Even when they do, OIT is only treated as a global transparent rendering solution, separate from traditional sorting-based transparent methods. Therefore, there is a lack of readily available reference cases or best practices for using Weighted Blended objects alongside ordinary alpha blend objects.

We aim to support Weighted Blended OIT and alpha blend objects simultaneously to some extent. However, the unique aspect of Weighted Blended OIT is its inherently multi-pass nature (at least requiring Pass A: rendering OIT objects to an accumulation/revealage buffer; Pass B: compositing to the main framebuffer via a full-screen pass). This differs from the standard forward rendering process. When mixing Weighted Blended OIT objects with ordinary alpha blend objects, frequent GPU rendering state switches are difficult to avoid:

- Rendering OIT objects requires switching to an OIT-specific render pass (binding the accumulation/revealage buffers).
- Rendering ordinary alpha objects requires switching back to the main framebuffer's render pass.

If these two types of objects appear alternately, it will trigger expensive render pass switches.

A reasonable idea is: since Weighted Blended OIT objects do not participate in depth sorting themselves, there is no need to handle interleaving scenarios between them and ordinary alpha blend objects. Therefore, all OIT objects can be processed uniformly at a certain stage of the rendering pipeline, thereby avoiding the complexity introduced by their multi-pass nature—that is, treating OIT as a completely independent rendering stage, decoupled from the rendering process of other objects.

Based on this, categorizing Weighted Blended OIT objects as a separate type and assigning them an independent rendering stage is a natural implementation. Although some users might desire the flexibility to manually specify the rendering order to achieve interleaved rendering of OIT and ordinary transparent objects, the value of this flexibility is likely outweighed by the performance cost and implementation complexity it introduces. Frequent render pass switches are not only prone to unintentional misuse by users, leading to performance degradation, but mixing Weighted Blended objects with ordinary alpha blend objects might also cause unexpected rendering issues. Furthermore, this would impose constraints on the engine's feature expansion and maintenance. For example, adding support for two-pass rendering of double-sided transparent objects would necessitate additional handling for compatibility with Weighted Blended OIT objects, introducing unnecessary complexity.

**I believe this design choice is essentially an engineering trade-off. The core of real-time rendering technology lies in finding a balance between visual realism, performance cost, and implementation complexity. This is also the fundamental distinction between modern real-time rendering and offline rendering.** While the basic laws of physics are inherently simple—light propagation, energy conservation, laws of reflection and refraction, all expressible precisely in mathematics—offline rendering (e.g., path tracing, Monte Carlo integration) tends to follow these physical laws directly, approximating the real world through extensive sampling and computation. In contrast, real-time rendering heavily employs approximations and simplifications (e.g., BRDF approximations, precomputed IBL) to achieve acceptable visual quality within the constraints of limited performance budgets and engineering complexity. Handling Weighted Blended OIT as a separate stage adheres to this philosophy.

A similar trade-off is evident in the handling of transmissive light in #974. For transmissive objects, screen-space refraction (SSR) is needed to gather information about the scene behind them.
I initially considered generating a separate scene framebuffer for each transmissive object, but this would be too expensive. Each transmissive object would require re-rendering the entire scene, which is impractical for real-time rendering.
Subsequently, I thought about utilizing transparency sorting and layer-by-layer accumulation (hence the most significant change initially made to the Pygfx renderer in this PR was the requirement for transparency sorting). The result after drawing each object could serve as the background for the next object, thus avoiding the costly overhead of rendering the scene multiple times. However, since the GPU does not allow simultaneous reading and writing to the same texture, double buffering (ping-pong buffers) or an intermediate copy mechanism would be required. This would still cause render pass interruptions. When there are multiple transmissive objects in the scene, although it wouldn't require extra re-rendering of the scene, frequent render pass interruptions and switches would still lead to performance degradation (This is the same as the problem with weighted blended objects.). Moreover, this approach still cannot handle cases where transmissive objects intersect or self-occlude.

Therefore, I tried looking at implementations in mature engines, and the final answer was straightforward. The methods adopted by the reference implementation of the glTF 2.0 Transmission extension, as well as engines like Three.js, Babylon.js, and Unity HDRP, are all very "simple and direct": after all opaque objects are rendered, a copy of the background color buffer is generated, and all transmissive objects share this buffer for sampling. Although this scheme cannot handle the superposition of multiple layers of transmissive objects, it is simple, efficient, and practical.

I believe this is a simple trade-off, and one that makes sense. Firstly, in the vast majority of scenarios, there won't be overlapping and superposition of multiple transmissive objects. Furthermore, adopting this fixed strategy also inversely requires scene designers and technical artists to optimize scene design accordingly, rather than abusing the feature. Content creators inherently need to optimize asset structure and scene layout within the engine's capabilities to achieve the best visual results.


All these engines divide the rendering process into multiple stages—at least an opaque stage and a transparent stage—and perform specific tasks between these stages based on logical requirements. Generating the SSR framebuffer is one such task. Some global illumination, AO, volumetric fog, and even certain post-processing effects also need to be inserted between the opaque and transparent stages. Unity even exposes injection points between these stages to let users plug in custom logic or passes.

panxinmiao commented:
Sorry for post such a “long essay” without properly dividing it into paragraphs and section titles, 😅 . I’ll add some examples and supporting facts to certain arguments and viewpoints in the text when I have more time.

panxinmiao commented:
https://docs.unity3d.com/Packages/com.unity.render-pipelines.high-definition%4017.0/manual/rendering-execution-order.html

This documentation clearly outlines Unity’s rendering pipeline stages, including the fundamental opaque and transparent phases, and further explains the exact timing of tasks executed within or between these stages.

I haven’t found detailed documentation on how Unity assigns renderable objects to different rendering stages, but I believe the division is based on whether the object’s renderQueue is greater than 2500 (opaque stage or transparent stage).

https://docs.unity3d.com/ScriptReference/Rendering.RenderQueue.html
The documentation emphasizes that users should “**Use the render queue section that matches the object type**”, and it also specifies that **2500 is the last render queue treated as opaque**.
